apps/prairielearn/elements/pl-number-input/pl-number-input.py

```diff
@@ -372,68 +410,78 @@ def grade(element_html, data):
-    elif comparison == "decdig":
-        digits = pl.get_integer_attrib(element, "digits", DIGITS_DEFAULT)
-        correct = pl.is_correct_scalar_dd(a_sub, a_tru, digits)
-    else:
-        raise ValueError('method of comparison "%s" is not valid' % comparison)
+    comparison = pl.get_enum_attrib(
+        element, "comparison", ComparisonType, COMPARISON_DEFAULT
+    )
 
-    if correct:
-        data["partial_scores"][name] = {"score": 1, "weight": weight}
-    else:
-        data["partial_scores"][name] = {"score": 0, "weight": weight}
+    def grade_function(a_sub: Union[str, dict[str, Any]]) -> tuple[bool, None]:
+        nonlocal a_tru
+        # If submitted answer is in a format generated by pl.to_json, convert it
+        # back to a standard type (otherwise, do nothing)
+        a_sub_parsed: Any = pl.from_json(a_sub)
+
+        # Cast both submitted and true answers as np.float64, because...
+        #
+        #   If the method of comparison is relabs (i.e., using relative and
+        #   absolute tolerance) then np.allclose is applied to check if the
+        #   submitted and true answers are the same. If either answer is an
+        #   integer outside the range of int64...
+        #
+        #       https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html
+        #
+        #   ...then numpy throws this error:
+        #
+        #       TypeError: ufunc 'isfinite' not supported for the input types, and
+        #       the inputs could not be safely coerced to any supported types
+        #       according to the casting rule ''safe''
+        #
+        #   Casting as np.float64 avoids this error. This is reasonable in any case,
+        #   because <pl-number-input> accepts double-precision floats, not ints.
+        #
+        if np.iscomplexobj(a_sub_parsed) or np.iscomplexobj(a_tru):
+            a_sub_converted = np.complex128(a_sub_parsed)
+            a_tru_converted = np.complex128(a_tru)
+        else:
+            a_sub_converted = np.float64(a_sub_parsed)
+            a_tru_converted = np.float64(a_tru)
+
+        # Compare submitted answer with true answer
+        if comparison is ComparisonType.RELABS:
+            rtol = pl.get_float_attrib(element, "rtol", RTOL_DEFAULT)
+            atol = pl.get_float_attrib(element, "atol", ATOL_DEFAULT)
+
+            a_sub_precision = get_string_precision(str(a_sub))
+            if not math.isclose(a_sub_precision, rtol, rel_tol=rtol):
+                feedback = ANSWER_INSUFFICIENT_PRECISION_WARNING
+
+            return (
+                pl.is_correct_scalar_ra(a_sub_converted, a_tru_converted, rtol, atol),
+                feedback,
+            )
+        elif comparison is ComparisonType.SIGFIG:
+            digits = pl.get_integer_attrib(element, "digits", DIGITS_DEFAULT)
+            return (
+                pl.is_correct_scalar_sf(a_sub_converted, a_tru_converted, digits),
+                None,
```

Comment:
@Frozemint So right now in the case of a correct answer, the following gets displayed:
![image](https://github.com/PrairieLearn/PrairieLearn/assets/1345068/cddf329e-789b-46f6-8fc9-ff6bc54eacf0)
This is effectively feedback that I'd rather have returned here (instead of stored in a popover). In the mustache template, can you remove this part of the submission panel and have that be handled as part of the grading logic here?

Reply:
So something that looks like this? 
![image](https://github.com/PrairieLearn/PrairieLearn/assets/11191061/3f859c16-c856-4e86-8f0c-fabb50bf86fd)
Right now the `the correct answer used for grading` text is not part of the grading logic, it is hardcoded text in the mustache, what I did was move the text from a popover into its own feedback span instead.

Reply:
> So something that looks like this?

Exactly! Looks great. Moving the correct answer feedback into the grading logic is the way to go.

Reply:
Unresolving this since this change got lost in the merge.

Reply:
Fixed again in [bfb518d](https://github.com/PrairieLearn/PrairieLearn/pull/7934/commits/bfb518dc636477ea6722f16869a75d7447386577).